core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala

```diff
@@ -1331,6 +1336,72 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     assert(store.getSingle("a1").isDefined, "a1 was not in store")
     assert(store.getSingle("a3").isDefined, "a3 was not in store")
   }
+
+  test("SPARK-13328: refresh block locations (fetch should fail after hitting a threshold)") {
+    val mockBlockTransferService = new MockBlockTransferService(maxFailuresBeforeLocationRefresh)
+    store = makeBlockManager(8000, "executor1")
+
+    store.putSingle("item", 999L, StorageLevel.MEMORY_ONLY, tellMaster = true)
+
+    intercept[BlockFetchException] {
+      store.doGetRemote("item", mockBlockTransferService, asBlockResult = false)
+      .asInstanceOf[Option[ByteBuffer]]
+    }
+  }
+
+  test("SPARK-13328: refresh block locations (fetch should succeed after location refresh)") {
+    val mockBlockManagerMaster = mock(classOf[BlockManagerMaster])
+    val mockBlockTransferService = new MockBlockTransferService(maxFailuresBeforeLocationRefresh)
+
+    // make sure we have more than maxFailuresBeforeLocationRefresh locations
+    // so that we have a chance to do location refresh
+    val blockManagerIds = new ArrayBuffer[BlockManagerId]()
+    for (i <- 0 to maxFailuresBeforeLocationRefresh) {
+      blockManagerIds += BlockManagerId(s"id-$i", s"host-$i", i + 1)
+    }
+
+    when(mockBlockManagerMaster.getLocations(mc.any[BlockId])).thenReturn(blockManagerIds.toSeq)
+
+    store = makeBlockManager(8000, "executor1", mockBlockManagerMaster)
+
+    val block = store.doGetRemote("item", mockBlockTransferService, asBlockResult = false)
+      .asInstanceOf[Option[ByteBuffer]]
+    assert(block.isDefined)
+    verify(mockBlockManagerMaster, times(2)).getLocations("item")
+  }
+
+  class MockBlockTransferService(val maxFailures: Int) extends BlockTransferService {
+    var numCalls = 0
+
+    override def init(blockDataManager: BlockDataManager): Unit = {}
+
+    override def fetchBlocks(host: String, port: Int, execId: String, blockIds: Array[String],
+                             listener: BlockFetchingListener): Unit = {
+      listener.onBlockFetchSuccess("mockBlockId", new NioManagedBuffer(ByteBuffer.allocate(1)))
+    }
+
+    override def close(): Unit = {}
+
+    override def hostName: String = { "MockBlockTransferService::host" }
+
+    override def port: Int = { -1 }
+
+    override def uploadBlock(hostname: String, port: Int, execId: String,
+                             blockId: BlockId, blockData: ManagedBuffer,
+                             level: StorageLevel): Future[Unit] = {
```

Comment:
on all of these methods, the style for multiline arg defs is each arg on its own line, indented 4 spaces, like:

```
override def uploadBlock(
    hostname: String,
    port: Int,
    execId: String,
    blockId: BlockId,
    blockData: ManagedBuffer,
    level: StorageLevel): Future[Unit] = {
```